File: Luogu/P_2986_USACO_10_MAR_Great_Cow_Gathering_G.cpp
#include <bits/stdc++.h>
using LL = long long;
using namespace std;
const LL INF = 1e18;
#define endl '\n'
#define PII pair<int, int>

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    LL n, u, v, w, tot = 0;
    cin >> n;
    vector<int> c(n);
    vector<vector<PII>> g(n);
    for (int i = 0; i < n; i++) {
        cin >> c[i];
        tot += c[i];
    }
    for (int i = 1; i < n; i++) {
        cin >> u >> v >> w;
        u--, v--;
        g[u].push_back({v, w});
        g[v].push_back({u, w});
    }

    vector<LL> sz(n), dep(n), f(n);
    auto dfs1 = [&](auto dfs1, int u, int fa) -> void {
        sz[u] = c[u];
        for (auto [v, w] : g[u]) {
            if (v == fa) continue;
            dep[v] = dep[u] + 1;
            dfs1(dfs1, v, u);
            sz[u] += sz[v];
            f[u] += f[v] + w * sz[v];
        }
    };
    dfs1(dfs1, 0, -1);

    LL ans = INF;
    auto dfs2 = [&](auto dfs2, int u, int fa) -> void {
        ans = min(ans, f[u]);
        for (auto [v, w] : g[u]) {
            if (v == fa) continue;
            f[v] = f[u] - sz[v] * w + (tot - sz[v]) * w;
            dfs2(dfs2, v, u);
        }
    };
    dfs2(dfs2, 0, -1);
    cout << ans << endl;
    return 0;
}

File: Luogu/P_2986_USACO_10_MAR_Great_Cow_Gathering_G.py
n = int(input())
c = [int(input()) for _ in range(n)]
tot = sum(c)

g = [[] for _ in range(n)]
for _ in range(n - 1):
    u, v, w = map(int, input().split())
    u, v = u - 1, v - 1
    g[u].append((v, w))
    g[v].append((u, w))

sz = [0] * n
dep = [0] * n
f = [0] * n
st = [(0, -1, True)]
while st:
    u, fa, flag = st.pop()
    if flag:
        sz[u] = c[u]
        st.append((u, fa, False))
        for v, w in g[u]:
            if v == fa:
                continue
            dep[v] = dep[u] + 1
            st.append((v, u, True))
    else:
        for v, w in g[u]:
            if v == fa:
                continue
            sz[u] += sz[v]
            f[u] += f[v] + w * sz[v]

ans = float('inf')
st = [(0, -1)]
while st:
    u, fa = st.pop()
    ans = min(ans, f[u])
    for v, w in g[u]:
        if v == fa:
            continue
        f[v] = f[u] - sz[v] * w + (tot - sz[v]) * w
        st.append((v, u))

print(ans)